Integrate PeeringState into crimson, fix related bugs

athanatos opened the pull request.

tchaikov commented:
retest this please.

tchaikov commented on src/crimson/osd/osd.h:
```diff
@@ -137,6 +145,61 @@ class OSD : public ceph::net::Dispatcher,
+      spg_t,
+      std::unique_ptr<PGPeeringEvent>>>;
+    return seastar::do_with(
+      std::make_tuple(PeeringState::PeeringCtx{}, std::move(f),
```
could also put
```c++
return seastar::do_with(PeeringState::PeeringCtx{}, F(std::move(f)), G(std::move(on_missing_pg)), 
                        [this, &c](auto& rtx, auto& f, auto& on_missing_pg) {
 // ...
}
```

tchaikov commented on src/crimson/osd/pg.cc:
```diff
@@ -43,933 +41,211 @@ namespace {
-  // if there are multiples, prefer
-  //  - a longer tail, if it brings another peer into log contiguity
-  //  - the current primary
-  struct is_good {
```
instead of completely dropping the code, i'd like at least port part of it to the `PeeringState.cc` for better readability once this change is merged.

athanatos replied:
Sure, this is probably easier to read than what's in there.

tchaikov commented on src/crimson/osd/osd.cc:
```diff
@@ -5,6 +5,7 @@
 #include <boost/smart_ptr/make_local_shared.hpp>
 #include <fmt/format.h>
 #include <fmt/ostream.h>
+#include <seastar/core/sleep.hh>
```
is `seastar::sleep()` or its friends used here?

athanatos replied:
Nope, I think I added it for debugging, will remove.

tchaikov commented:
retest this please

rzarzynski commented on src/crimson/os/cyan_store.cc:
```diff
@@ -287,19 +313,78 @@ seastar::future<> CyanStore::do_transaction(CollectionRef ch,
+        ghobject_t oid = i.get_oid(op->oid);
+        map<string, bufferlist> aset;
+        i.decode_attrset(aset);
+        r = _omap_set_values(cid, oid, aset);
```
Possibly there is a chance for `std::moving` the `aset` to `_omap_set_values()`.

athanatos replied:
_omap_set_values is taking a const &, so probably not important.

rzarzynski replied:
`_omap_set_values()` consumes the container to store its pairs in `Object::omap`, so it could take just `std::map<std::string, ceph::bufferlist> aset`. Still, I agree it can be unimportant – no evidence from e.g. profiling we need to optimize this.

athanatos replied:
Ok, I'll switch it to consume the container.

rzarzynski commented on src/crimson/os/cyan_store.h:
```diff
@@ -21,7 +22,11 @@ class Transaction;
 
 // a just-enough store for reading/writing the superblock
 class CyanStore {
+  constexpr static int MAX_KEYS_PER_OMAP_GET_CALL = 32;
```
nit: `int`.

athanatos replied:
Are you saying it should be unsigned?  Or a shorter int type?

rzarzynski replied:
Apologizes for being not specific. I think we don't need sign here, so an unsigned would fit better. `size_t` instead?

tchaikov replied:
probably unsigned would be better? as `size_t` is for size.

athanatos replied:
Switching to unsigned.

athanatos commented on src/osd/PeeringState.h:
```diff
@@ -1514,15 +1514,16 @@ class PeeringState : public MissingLoc::MappingInfo {
     pg_info_t &&info_from_disk,
     PastIntervals &&past_intervals_from_disk,
-    F &&pg_log_init) {
+    F &&pg_log_init) -> decltype(pg_log_init(pg_log)) {
```
Ok

athanatos commented on src/crimson/os/cyan_store.cc:
```diff
@@ -386,7 +386,16 @@ seastar::future<> CyanStore::do_transaction(CollectionRef ch,
     ceph_assert(r == 0);
   }
-  return seastar::now();
+  return seastar::now().then([t=std::move(t)]() mutable {
```
Well, it needs to return a future<>, but I can certainly call the completions inline and then return seastar::now().  I suppose the only thing the above does is give the scheduler an opportunity to preempt if it wants to.  I'll just call them inline.

tchaikov replied:
oh, right! yeah, i meant to remove the leading first `seastar::now()` call.

athanatos commented on src/crimson/os/cyan_store.cc:
```diff
@@ -232,30 +232,56 @@ CyanStore::omap_get_values(CollectionRef c,
+    values.insert(*i);
+  }
+  return seastar::make_ready_future<bool, omap_values_t>(
+    true, omap_values_t());
```
Ah, yes.

athanatos commented on src/crimson/net/SocketMessenger.h:
```diff
@@ -44,6 +45,9 @@ class SocketMessenger final : public Messenger, public seastar::peering_sharded_
   void do_bind(const entity_addrvec_t& addr);
+
+  bool started = false;
+  seastar::shared_promise<> accepting_complete;
```
I don't need to maintain a count, nor do I need to block enter, I just need to signal when the accept loop terminates.

athanatos commented on src/crimson/os/cyan_store.h:
```diff
@@ -101,6 +116,23 @@ class CyanStore {
+  int _omap_set_values(
+    const coll_t& cid,
+    const ghobject_t& oid,
+    const map<string, bufferlist> &aset);
```
Yep, thanks.

athanatos commented on src/osd/osd_types.h:
```diff
@@ -3014,6 +3015,11 @@ class OSDMap;
  */
 class PastIntervals {
+#ifdef WITH_SEASTAR
+  using OSDMapRef = boost::local_shared_ptr<const OSDMap>;
```
It's what crimson is already using, I'm inclined to change that in a future pr.

athanatos commented on src/osd/PGLog.h:
```diff
@@ -1570,4 +1576,136 @@ struct PGLog : DoutPrefixProvider {
+
+    map<eversion_t, hobject_t> divergent_priors;
+    bool must_rebuild = false;
+    list<pg_log_entry_t> entries;
```
Sure

athanatos commented on src/osd/OSDMap.h:
```diff
@@ -1504,7 +1505,12 @@ class OSDMap {
 
-typedef std::shared_ptr<const OSDMap> OSDMapRef;
+#ifdef WITH_SEASTAR
+using OSDMapRef = boost::local_shared_ptr<const OSDMap>;
```
It's what crimson is already using, I'm inclined to change that in a future pr.

rzarzynski replied:
OK, `TODO` comment is fine.

rzarzynski replied:
> It's what crimson is already using

Yeah, I guess e.g. `SharedLRU` needs `seastar::shared_ptr<...>` because of deleter.

tchaikov commented on src/crimson/osd/osd.cc:
```diff
@@ -854,24 +1056,95 @@ seastar::future<> OSD::consume_map(epoch_t epoch)
+OSD::get_pg(
+  spg_t pgid,
+  epoch_t epoch,
+  std::unique_ptr<PGCreateInfo> info) {
```
nit, could you move the `{` down to its own line? just so my buggy builtin parser can work better. =)

athanatos replied:
Whoops, yeah, that's the convention for .cc files.

tchaikov commented on src/crimson/osd/osdmap_service.h:
```diff
@@ -11,7 +11,7 @@ class OSDMap;
 class OSDMapService {
 public:
-  using cached_map_t = boost::local_shared_ptr<OSDMap>;
+  using cached_map_t = boost::local_shared_ptr<const OSDMap>;
```
makes sense!

rzarzynski replied:
:+1:

tchaikov commented on src/crimson/osd/pg.h:
```diff
@@ -36,99 +38,398 @@ namespace ceph::os {
+     ec_profile_t profile);
+
+  // EpochSource
+  epoch_t get_osdmap_epoch() const override final {
```
nit, IMHO, `final` would suffice. and i think https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c128-virtual-functions-should-specify-exactly-one-of-virtual-override-or-final explains the rationale well.

athanatos replied:
Ok, will switch them all.

tchaikov commented on src/crimson/osd/shard_services.cc:
```diff
@@ -0,0 +1,223 @@
+seastar::future<> ShardServices::dispatch_context_messages(
+  PeeringState::PeeringCtx &ctx)
+{
+  auto ret = seastar::do_with(PeeringState::PeeringCtx{ctx}, [this](auto& todo) {
```
can we capture-by-move `ctx.notify_list`, `ctx.query_map` and `ctx.info_map` instead of copying them?

athanatos replied:
I think I did for the bits below, but I'll update the initial do_with to capture-by-move PeeringCtx.

tchaikov commented on src/crimson/osd/shard_services.cc:
```diff
@@ -0,0 +1,223 @@
+  logger().debug("{}: {}", __func__, pg_temp_wanted);
+  MOSDPGTemp *ms[2] = {nullptr, nullptr};
+  for (auto& [pgid, pg_temp] : pg_temp_wanted) {
+    auto& m = ms[pg_temp.forced];
```
nit, `auto m` would suffice.

tchaikov commented on src/crimson/osd/shard_services.cc:
```diff
@@ -0,0 +1,223 @@
+  for (auto& [pgid, pg_temp] : pg_temp_wanted) {
+    auto& m = ms[pg_temp.forced];
+    if (!m) {
+      m = new MOSDPGTemp(osdmap->get_epoch());
```
i'd suggest use `make_message< MOSDPGTemp>(osdmap->get_epoch())` instead, because we always tend to forget that the ctor of `intrusive_ptr(T * p, bool add_ref = true);` increments the ref by default.

tchaikov commented on src/crimson/osd/shard_services.cc:
```diff
@@ -0,0 +1,223 @@
+  if (pg_temp_wanted.empty())
+    return;
+  logger().debug("{}: {}", __func__, pg_temp_wanted);
+  MOSDPGTemp *ms[2] = {nullptr, nullptr};
```
might want to put
```c++
boost::intrusive_ptr<MOSDPGTemp>[2];
```
see below for the rationale .

tchaikov commented on src/crimson/osd/shard_services.h:
```diff
@@ -0,0 +1,129 @@
+
+#include "crimson/os/cyan_store.h"
+#include "crimson/osd/osdmap_service.h"
+#include "crimson/mgr/client.h"
```
this is not necessary.

tchaikov commented on src/crimson/osd/shard_services.h:
```diff
@@ -0,0 +1,129 @@
+#include "crimson/os/cyan_store.h"
+#include "crimson/osd/osdmap_service.h"
+#include "crimson/mgr/client.h"
+#include "crimson/mon/MonClient.h"
```
neither is this. as we can forward declare `ceph::mon::Client` as well.

tchaikov commented on src/crimson/osd/shard_services.h:
```diff
@@ -0,0 +1,129 @@
+#include "crimson/osd/osdmap_service.h"
+#include "crimson/mgr/client.h"
+#include "crimson/mon/MonClient.h"
+#include "crimson/net/Messenger.h"
```
ditto.

athanatos replied:
I can actually get rid of all of these by moving PeeringCtx and Collection out of their respective classes, will do.

tchaikov commented on src/crimson/osd/shard_services.h:
```diff
@@ -0,0 +1,129 @@
+  // TODO: hook into map processing and some kind of heartbeat/peering
+  // message processing
+  struct pg_temp_t {
+    vector<int> acting;
```
`std::vector<>`.

tchaikov commented on src/crimson/osd/shard_services.h:
```diff
@@ -0,0 +1,129 @@
+#include "crimson/mgr/client.h"
+#include "crimson/mon/MonClient.h"
+#include "crimson/net/Messenger.h"
+#include "crimson/net/Connection.h"
```
we don't need the declaration of `Connection` here, do we?

tchaikov commented on src/crimson/osd/shard_services.h:
```diff
@@ -0,0 +1,129 @@
+
+#include <seastar/core/future.hh>
+
+#include "crimson/os/cyan_store.h"
```
probably we can forward declare CyanStore?

athanatos commented on src/crimson/mon/MonClient.h:
```diff
@@ -91,6 +90,10 @@ class Client : public ceph::net::Dispatcher,
   bool sub_want_increment(const std::string& what, version_t start, unsigned flags);
   seastar::future<> renew_subs();
 
+  MonMap &get_monmap_ref() {
```
Will remove.

athanatos commented on src/crimson/osd/osd.cc:
```diff
@@ -526,6 +538,167 @@ seastar::future<> OSD::osdmap_subscribe(version_t epoch, bool force_request)
+	  pg->handle_activate_map(rctx);
+
+	  logger().info("{} new pg {}", __func__, *pg);
+          pgs.emplace(info->pgid, pg);
```
I don't think it's being moved away there, pg (Ref<PG>) is being copied.

tchaikov replied:
yeah, you are correct. it's different than `pgs.emplace(info->pgid, std::move(pg));`

tchaikov commented on src/osd/PeeringState.h:
```diff
@@ -54,6 +54,59 @@ struct PGPool {
+  }
+
+  void accept_buffered_messages(BufferedRecoveryMessages &m) {
+    for (map<int, map<spg_t, pg_query_t> >::iterator i = m.query_map.begin();
```
can we use ranged loop and structured binding here?

tchaikov commented on src/osd/PeeringState.h:
```diff
@@ -54,6 +54,59 @@ struct PGPool {
+	omap[j->first] = j->second;
+      }
+    }
+    for (map<int, vector<pair<pg_notify_t, PastIntervals> > >::iterator i
```
ditto

tchaikov commented on src/osd/PeeringState.h:
```diff
@@ -54,6 +54,59 @@ struct PGPool {
+	 i != m.query_map.end();
+	 ++i) {
+      map<spg_t, pg_query_t> &omap = query_map[i->first];
+      for (map<spg_t, pg_query_t>::iterator j = i->second.begin();
```
ditto.

tchaikov commented on src/osd/PeeringState.h:
```diff
@@ -54,6 +54,59 @@ struct PGPool {
+      ovec.reserve(ovec.size() + i->second.size());
+      ovec.insert(ovec.end(), i->second.begin(), i->second.end());
+    }
+    for (map<int, vector<pair<pg_notify_t, PastIntervals> > >::iterator i
```
ditto.

athanatos replied:
Yeah, that actually made it far clearer.

athanatos commented:
I've been using

make -j9 crimson-osd && ( pkill -9 'crimson|ceph' || true ) && CEPH_NUM_MDS=0 ../src/vstart.sh --crimson --nodaemon -n -d --redirect-output --msgr1 --osd-args '--debug-osd=20'

to run the osds.  Try that?  If it doesn't work, can you post what you tried to do?

athanatos commented:
@tchaikov Oh, actually:

for i in 0 1 2; do ./bin/ceph osd crush add osd.$i 1 host=myhost; done;  ./bin/ceph osd crush move myhost root=default && ./bin/ceph osd pool create foo3 16 16

Classic osd assigns itself to the right host/root on startup -- crimson hasn't learned that trick yet.  You need to explicitly set the crush hierarchy for the pool to be assigned right.

athanatos commented:
@tchaikov I updated the 'crimson/mon/MonClient: handle mon client connection race' to correctly handle the -EAGAIN.  I don't think it was causing an actual problem, but I missed the do_auth retry loop in the original version.

I also added patches to make remove behave correctly.  With that patch, I'm able to get 3 osds to peer, do writes, reads, and removes with msgr1 and a single osd to do the same operations with msgr 1 or 2.  Let me know if you're still having trouble, I'll kick off a qa run in a few hours when it builds.

athanatos commented:
Added a few misc bug fixes with reading memstore state left from classic -- I hadn't tested starting a crimson osd after a classic osd in a while and evidently I broke it.

athanatos commented:
retest this please

tchaikov commented:
@athanatos i tested with following combinations of 

- 3 OSDs, with msg1
- 3 OSDs, with msg2
- 1 OSD, with msg1
- 1 OSD, with msg2

your changeset works great!

the only issue i ran into was that if i

1. create a vstart cluster with a single classic OSD
1. create a pool
1. stop the classic OSD
1. start the crimson OSD

crimson-osd crashes. because we call `try_flush_or_schedule_async()` when handling a peering event, and this method can emit yet another event. and `PeeringState::handle_event()` is not reentrant -- `PeeringState::start_handle()` has `ceph_assert(!rctx)`. but classic OSD does not suffer from it, because its implementation of `PG::try_flush_or_schedule_async()` schedules `PeeringState::IntervalFlush()` using the objectstore's finisher thread.

probably, in a follow-up PR, we could revise `PG::do_peering_event(const boost::statechart::event_base &, PeeringCtx)` to schedule the new peering event until the the peering state finish handling existing one if any, something like:

1. wrap `peering_state.handle_event(...)` in a continuation
1. chain it onto a `shared_future` which is in turn a member variable of `PG`
1. and replace the `shared_future` with the new one. 

what do you think?

tchaikov commented on src/crimson/osd/pg.h:
```diff
@@ -36,99 +38,398 @@ namespace ceph::os {
+    shard_services.send_pg_created(pgid);
+  }
+
+  bool try_flush_or_schedule_async() final {
```
might want to move the implementation of this method to .cc file.

tchaikov commented on src/crimson/osd/pg.h:
```diff
@@ -36,99 +38,398 @@ namespace ceph::os {
-			    int new_up_primary,
-			    const std::vector<int>& new_acting,
-			    int new_acting_primary);
+
```
could you remove unnecessary empty lines?

tchaikov commented on src/crimson/osd/pg.h:
```diff
@@ -36,99 +38,398 @@ namespace ceph::os {
-  seastar::future<> handle_activate_map();
-  seastar::future<> share_pg_info();
-  void reply_pg_query(const MQuery& query, recovery::Context* ctx);
+  void do_peering_event(
```
might want to make this method private.

athanatos replied:
Called from osd.

rzarzynski commented on src/crimson/osd/osd.h:
```diff
@@ -137,6 +146,61 @@ class OSD : public ceph::net::Dispatcher,
+    const C &c,
+    F &&f,
+    G &&on_missing_pg) {
+    using mapped_type = decltype(*(c.begin()));
```
It appears this template deals with regular containers as:

* `MOSDPGQuery::pg_list` is `std::map<spg_t, pg_query_t>`,
* `MOSDPGCreate2::pgs` – `map<spg_t,pair<epoch_t,utime_t>>` while
* `MOSDPGNotify::get_pg_list()` returns `std::vector<std::pair<pg_notify_t,PastIntervals>>` .

Maybe instead of the `decltype` we could use something like:

```cpp
      using mapped_type = typename C::value_type;
```

rzarzynski commented on src/crimson/osd/osd.h:
```diff
@@ -137,6 +146,61 @@ class OSD : public ceph::net::Dispatcher,
+    return handle_batch_pg_message_with_missing_handler(
+      c,
+      std::move(f),
+      [](decltype(*(c.begin())), PeeringCtx &){});
```
Maybe `typename C::value_type`?

rzarzynski commented on src/crimson/osd/pg.h:
```diff
@@ -36,99 +38,398 @@ namespace ceph::os {
+    bool dirty_big_info,
+    bool need_write_epoch,
+    ObjectStore::Transaction &t) final {
+    map<string,bufferlist> km;
```
nit: not sure we should assume `using namespace std`.

rzarzynski commented on src/crimson/osd/pg.h:
```diff
@@ -36,99 +38,398 @@ namespace ceph::os {
+    int role,
+    const vector<int>& up,
+    int up_primary,
+    const vector<int>& acting,
```
nit: `std`.

rzarzynski commented on src/crimson/osd/shard_services.cc:
```diff
@@ -0,0 +1,230 @@
+  for (auto& [pgid, pg_temp] : pg_temp_wanted) {
+    auto m = ms[pg_temp.forced];
+    if (!m) {
+      m = make_message< MOSDPGTemp >(osdmap->get_epoch());
```
nit: do we need the extra whitechars here? In this file the convention is rather `make_message<MOSDPGInfo>`.

rzarzynski commented on src/crimson/osd/shard_services.cc:
```diff
@@ -0,0 +1,230 @@
+    }
+    m->pg_temp.emplace(pgid, pg_temp.acting);
+  }
+  for (auto m : ms) {
```
nit: this imposes extra ref-counting but I'm really not sure we need to worry in this particular case.

athanatos commented:
qa run showed that I apparently subtly broke the may_include_deletes flag, so I'll address the above comments, rerun, and hopefully merge on Monday.

jdurgin commented:
Note that recovery_deletes is always enabled in the osdmap since luminous (which all upgrades need to pass through), so the only time we'd need the !may_include_deletes code would be for handling older on-disk state for an osd that didn't boot since prior to luminous.

athanatos commented:
@jdurgin Yeah, one of the patches in this series simply looks at the osdmap in init and sets it immediately rather than waiting for the next map -- I think that behavior was subtly broken anyway, though I suppose a newly created pg would never have objects to delete.  I missed setting the rebuilt_... flag so the corresponding omap key didn't get written out.

athanatos commented:
@tchaikov Looks like I'm going to change up the peering event scheduling again with the operation ordering PR -- I'll create an appropriate way to handle it there.  For now, I'll switch it to just lie.

athanatos commented:
@jdurgin Actually, could have a look at the last patch in the series if you've got time?

athanatos commented:
retest this please

jdurgin commented:
The last patch looks fine to me, the only semantic change looks to be PeeringState::init(), which makes sense. I think we could set missing_may_contain_deletes() unconditionally in octopus, but it doesn't make much difference to the amount of code.
